# Other species not in the list
Here, we provide a tutorial for the species not included in the given list but in the basic Homer installation. You can choose one genome version from the following table based on your data.
| **Genome Version**  | **Release** | **Description**                                           |
|---------------------|-------------|-----------------------------------------------------------|
| tair10              | v6.0        | Arabidopsis genome and annotation (tair10)                |
| dm6                 | v7.0        | Fly genome and annotation for UCSC dm6                    |
| xenTro3             | v6.4        | Frog genome and annotation for UCSC xenTro3               |
| danRer11            | v7.0        | Zebrafish genome and annotation for UCSC danRer11         |
| rn7                 | v7.0        | Rat genome and annotation for UCSC rn7                    |
| canFam3             | v6.4        | Dog genome and annotation for UCSC canFam3                |
| anoGam3             | v7.0        | Mosquito genome and annotation for UCSC anoGam3           |
| gorGor3             | v6.4        | Human genome and annotation for UCSC gorGor3              |
| strPur2             | v7.0        | Urchin genome and annotation for UCSC strPur2             |
| rn5                 | v6.4        | Rat genome and annotation for UCSC rn5                    |
| rheMac10            | v7.0        | Rhesus genome and annotation for UCSC rheMac10            |
| danRer7             | v6.4        | Zebrafish genome and annotation for UCSC danRer7          |
| canFam5             | v7.0        | Dog genome and annotation for UCSC canFam5                |
| gorGor6             | v7.0        | Human genome and annotation for UCSC gorGor6              |
| canFam6             | v7.0        | Dog genome and annotation for UCSC canFam6                |
| apiMel2             | v7.0        | Bee genome and annotation for UCSC apiMel2                |
| gorGor5             | v6.4        | Human genome and annotation for UCSC gorGor5              |
| panTro4             | v6.4        | Human genome and annotation for UCSC panTro4              |
| xenTro7             | v6.4        | Frog genome and annotation for UCSC xenTro7               |
| rheMac2             | v6.4        | Rhesus genome and annotation for UCSC rheMac2             |
| gorGor4             | v6.4        | Human genome and annotation for UCSC gorGor4              |
| panTro5             | v6.4        | Human genome and annotation for UCSC panTro5              |
| panPan2             | v6.4        | Human genome and annotation for UCSC panPan2              |
| patens.ASM242v1     | v5.10       | Patens genome and annotation (patens.ASM242v1)            |
| panTro6             | v7.0        | Human genome and annotation for UCSC panTro6              |
| AGPv3               | v5.10       | Corn genome and annotation (AGPv3)                        |
| xenTro9             | v6.4        | Frog genome and annotation for UCSC xenTro9               |
| papAnu2             | v6.4        | Human genome and annotation for UCSC papAnu2              |
| corn.AGPv3          | v5.10       | Corn genome and annotation (corn.AGPv3)                   |
| petMar2             | v6.4        | Lamprey genome and annotation for UCSC petMar2            |
| panPan1             | v6.4        | Human genome and annotation for UCSC panPan1              |
| fr3                 | v7.0        | Fugu genome and annotation for UCSC fr3                   |
| sacCer3             | v7.0        | Yeast genome and annotation for UCSC sacCer3              |
| mm9                 | v7.0        | Mouse genome and annotation for UCSC mm9                  |
| susScr3             | v6.4        | Pig genome and annotation for UCSC susScr3                |
| hg17                | v7.0        | Human genome and annotation for UCSC hg17                 |
| panTro3             | v6.4        | Human genome and annotation for UCSC panTro3              |
| tetNig2             | v7.0        | Fugu genome and annotation for UCSC tetNig2               |
| mm10                | v7.0        | Mouse genome and annotation for UCSC mm10                 |
| taeGut2             | v7.0        | Zebrafinch genome and annotation for UCSC taeGut2         |
| hg18                | v7.0        | Human genome and annotation for UCSC hg18                 |
| susScr11            | v7.0        | Pig genome and annotation for UCSC susScr11               |
| galGal5             | v6.4        | Chicken genome and annotation for UCSC galGal5            |
| hg38                | v7.0        | Human genome and annotation for UCSC hg38                 |
| mm8                 | v6.4        | Mouse genome and annotation for UCSC mm8                  |
| xenLae2             | v7.0        | Frog genome and annotation for UCSC xenLae2               |
| rn6                 | v7.0        | Rat genome and annotation for UCSC rn6                    |
| ce11                | v7.0        | Worm genome and annotation for UCSC ce11                  |
| dm3                 | v7.0        | Fly genome and annotation for UCSC dm3                    |
| rheMac3             | v6.4        | Rhesus genome and annotation for UCSC rheMac3             |
| rn4                 | v6.4        | Rat genome and annotation for UCSC rn4                    |
| rice.IRGSP1.0       | v5.10       | Rice genome and annotation (rice.IRGSP1.0)                |
| panPan3             | v7.0        | Human genome and annotation for UCSC panPan3              |
| galGal6             | v7.0        | Chicken genome and annotation for UCSC galGal6            |
| danRer10            | v7.0        | Zebrafish genome and annotation for UCSC danRer10         |
| ci2                 | v7.0        | Ciona genome and annotation for UCSC ci2                  |
| petMar3             | v7.0        | Lamprey genome and annotation for UCSC petMar3            |
| sacCer2             | v7.0        | Yeast genome and annotation for UCSC sacCer2              |
| hg19                | v7.0        | Human genome and annotation for UCSC hg19                 |
| xenTro10            | v7.0        | Frog genome and annotation for UCSC xenTro10              |
| rheMac8             | v6.4        | Rhesus genome and annotation for UCSC rheMac8             |
| mm39                | v7.0        | Mouse genome and annotation for UCSC mm39                 |
| ce10                | v7.0        | Worm genome and annotation for UCSC ce10                  |
| aplCal1             | v6.4        | Seahare genome and annotation for UCSC aplCal1            |
| xenTro2             | v6.4        | Frog genome and annotation for UCSC xenTro2               |
| papAnu4             | v7.0        | Human genome and annotation for UCSC papAnu4              |
| ce6                 | v7.0        | Worm genome and annotation for UCSC ce6                   |
| ci3                 | v7.0        | Ciona genome and annotation for UCSC ci3                  |
| apiMel3             | v7.0        | Bee genome and annotation for UCSC apiMel3                |
| anoGam1             | v7.0        | Mosquito genome and annotation for UCSC anoGam1           |
| galGal4             | v6.4        | Chicken genome and annotation for UCSC galGal4            |

## Prepare the input data
We take the sc data of mm10 as an example. The data is from the published paper (FOXA2 drives lineage plasticity and KIT pathway
activation in neuroendocrine prostate cancer).
The input data is the feature matrix from 10x sc-multiome data and Cell annotation/cell type label which includes: 
- Single-cell multiome data including matrix.mtx, features.tsv/features.txt, and barcodes.tsv/barcodes.txt

  If the input data is 10X h5 file or h5ad file from scanpy, please follow the instruction [h5/h5ad file as input](https://github.com/Durenlab/LINGER/blob/main/docs/h5_input.md) .
  
  Here, we provide sc-multiome data of mm10. We download the data using the shell command line.
```sh
mkdir -p data
cd data
wget --load-cookies /tmp/cookies.txt "https://drive.usercontent.google.com/download?export=download&confirm=$(wget --quiet --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://drive.usercontent.google.com/download?id=1PDOmtO2oL-YVxKQY26jL91SAFedDALA0'  -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\1\n/p')&id=1PDOmtO2oL-YVxKQY26jL91SAFedDALA0" -O mm10_data.tar.gz && rm -rf /tmp/cookies.txt
tar -xzvf mm10_data.tar.gz
mv mm10_data/* ./
cd ../
```
We provide the cell annotation as follows:
```sh
wget --load-cookies /tmp/cookies.txt "https://docs.google.com/uc?export=download&confirm=$(wget --quiet --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://docs.google.com/uc?export=download&id=1nFm5shjcDuDYhA8YGzAnYoYVQ_29_Yj4' -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\1\n/p')&id=1nFm5shjcDuDYhA8YGzAnYoYVQ_29_Yj4" -O mm10_label.txt && rm -rf /tmp/cookies.txt
mv mm10_label.txt data/label.txt
```
- Cell annotation/cell type label if you need the cell type-specific gene regulatory network (label.txt in our example).
  <div style="text-align: right">
  <img src="barcode_mm10.png" alt="Image" width="260">
  </div>  
- gtf file or the URL of the gtf file describing gene annotation, '*.gtf'
- PWM matrix file of motifs, 'all_motif.txt'
  <div style="text-align: right">
  <img src="PWM.jpg" alt="Image" width="400">
  </div>  
- Motif-TF match file, 'MotifMatch.txt', mapping motif and TFs
  <div style="text-align: right">
  <img src="motifmatch.png" alt="Image" width="300">
  </div>  
## LINGER 
### Install
```sh
conda create -n LINGER python==3.10.0
conda activate LINGER
pip install LingerGRN==1.97
conda install bioconda::bedtools #Requirement
```
### Install homer
Check whether homer is installed
```sh
which homer # run this in the command line
```
If homer is not installed, use Conda to install it
```sh
conda install bioconda::homer
```
#### Install genome
You can check the installed genome
```sh
dir=$(which homer)
dir_path=$(dirname "$dir")
ls $dir_path/../share/homer/data/genomes/  # this is the installed genomes, if no genome is installed, there will be an error 'No such file or directory'
```
If the genome is not installed, use the following shell script to install it.
```sh
genome='mm10'
perl $dir_path/../share/homer/configureHomer.pl -install $genome
```
For the following step, we run the code in python.
#### Input 
```python
GRNdir='path/to/current_dir/' # The dir we currently use; the sc data is in GRNdir/data/
gtf_file='*.gtf'# gtf, PWM matrix, and Motif-TF match file should be included in the GRNdir
PWM_file='all_motif.txt'
MotifMatch_file='MotifMatch.txt'
genome='mm10' # the genome of your data
outdir='/path/to/output/' #output dir
species='New'
```
#### Transfer the sc-multiome data to anndata  
We will transfer sc-multiome data to the anndata format and filter the cell barcode by the cell type label.
```python
import scanpy as sc
#set some figure parameters for nice display inside Jupiter notebooks.
sc.settings.set_figure_params(dpi=80, frameon=False, figsize=(5, 5), facecolor='white')
sc.settings.verbosity = 3  # verbosity: errors (0), warnings (1), info (2), hints (3)
sc.logging.print_header()
#results_file = "scRNA/pbmc10k.h5ad"
import scipy
import pandas as pd
matrix=scipy.io.mmread('data/matrix.mtx')
features=pd.read_csv('data/features.txt',sep='\t',header=None)
barcodes=pd.read_csv('data/barcodes.txt',sep='\t',header=None)
label=pd.read_csv('data/label.txt',sep='\t',header=0)
from LingerGRN.preprocess import *
adata_RNA,adata_ATAC=get_adata(matrix,features,barcodes,label)# adata_RNA and adata_ATAC are scRNA and scATAC
```
#### Remove low counts cells and genes
```python
import scanpy as sc
sc.pp.filter_cells(adata_RNA, min_genes=200)
sc.pp.filter_genes(adata_RNA, min_cells=3)
sc.pp.filter_cells(adata_ATAC, min_genes=200)
sc.pp.filter_genes(adata_ATAC, min_cells=3)
selected_barcode=list(set(adata_RNA.obs['barcode'].values)&set(adata_ATAC.obs['barcode'].values))
barcode_idx=pd.DataFrame(range(adata_RNA.shape[0]), index=adata_RNA.obs['barcode'].values)
adata_RNA = adata_RNA[barcode_idx.loc[selected_barcode][0]]
barcode_idx=pd.DataFrame(range(adata_ATAC.shape[0]), index=adata_ATAC.obs['barcode'].values)
adata_ATAC = adata_ATAC[barcode_idx.loc[selected_barcode][0]]
```
#### Generate the pseudo-bulk/metacell:
```python
from LingerGRN.pseudo_bulk import *
samplelist=list(set(adata_ATAC.obs['sample'].values)) # sample is generated from cell barcode 
tempsample=samplelist[0]
TG_pseudobulk=pd.DataFrame([])
RE_pseudobulk=pd.DataFrame([])
singlepseudobulk = (adata_RNA.obs['sample'].unique().shape[0]*adata_RNA.obs['sample'].unique().shape[0]>100)
for tempsample in samplelist:
    adata_RNAtemp=adata_RNA[adata_RNA.obs['sample']==tempsample]
    adata_ATACtemp=adata_ATAC[adata_ATAC.obs['sample']==tempsample]
    TG_pseudobulk_temp,RE_pseudobulk_temp=pseudo_bulk(adata_RNAtemp,adata_ATACtemp,singlepseudobulk)                
    TG_pseudobulk=pd.concat([TG_pseudobulk, TG_pseudobulk_temp], axis=1)
    RE_pseudobulk=pd.concat([RE_pseudobulk, RE_pseudobulk_temp], axis=1)
    RE_pseudobulk[RE_pseudobulk > 100] = 100

import os
if not os.path.exists('data/'):
    os.mkdir('data/')
adata_ATAC.write('data/adata_ATAC.h5ad')
adata_RNA.write('data/adata_RNA.h5ad')
TG_pseudobulk=TG_pseudobulk.fillna(0)
RE_pseudobulk=RE_pseudobulk.fillna(0)
pd.DataFrame(adata_ATAC.var['gene_ids']).to_csv('data/Peaks.txt',header=None,index=None)
TG_pseudobulk.to_csv('data/TG_pseudobulk.tsv')
RE_pseudobulk.to_csv('data/RE_pseudobulk.tsv')
```
### Training model
Overlap the region with general GRN:
```python
activef='ReLU' 
method='scNN'
import torch
import subprocess
import os
import LingerGRN.LINGER_tr as LINGER_tr
LINGER_tr.get_TSS_ensembl(genome,gtf_file)
LINGER_tr.get_TSS(GRNdir,genome,200000) # Here, 200000 represent the largest distance of regulatory element to the TG. Other distance is supported
LINGER_tr.RE_TG_dis(outdir)
```
Train for the LINGER model.
```python
import LingerGRN.LINGER_tr as LINGER_tr
activef='ReLU' # active function chose from 'ReLU','sigmoid','tanh'
LINGER_tr.training(GRNdir,method,outdir,activef,species)
```
### Cell population gene regulatory network
#### TF binding potential
The output is 'cell_population_TF_RE_binding.txt', a matrix of the TF-RE binding score.
```python
import LingerGRN.LL_net as LL_net
LL_net.TF_RE_binding(GRNdir,adata_RNA,adata_ATAC,genome,method,outdir)
```

#### *cis*-regulatory network
The output is 'cell_population_cis_regulatory.txt' with 3 columns: region, target gene, cis-regulatory score.
```python
LL_net.cis_reg(GRNdir,adata_RNA,adata_ATAC,genome,method,outdir)
```
#### *trans*-regulatory network
The output is 'cell_population_trans_regulatory.txt', a matrix of the trans-regulatory score.
```python
LL_net.trans_reg(GRNdir,method,outdir,genome)
```

### Cell type sepecific gene regulaory network
There are 2 options:
1. infer GRN for a specific cell type, which is in the label.txt;
```python
celltype='1' #use a string to assign your cell type
```
2. infer GRNs for all cell types.
```python
celltype='all'
```
Please make sure that 'all' is not a cell type in your data.
#### Motif matching
```python
command='paste data/Peaks.bed data/Peaks.txt > data/region.txt'
subprocess.run(command, shell=True)
import pandas as pd
command='findMotifsGenome.pl data/region.txt '+genome+' ./. -size given -find '+GRNdir+PWM_file+'> '+outdir+'MotifTarget.bed'
subprocess.run(command, shell=True)
```
#### TF binding potential
The output is 'cell_population_TF_RE_binding_*celltype*.txt', a matrix of the TF-RE binding potential.
```python
LL_net.cell_type_specific_TF_RE_binding(GRNdir,adata_RNA,adata_ATAC,genome,celltype,outdir,method)# different from the previous version
```

#### *cis*-regulatory network
The output is 'cell_type_specific_cis_regulatory_{*celltype*}.txt' with 3 columns: region, target gene, cis-regulatory score.
```python
LL_net.cell_type_specific_cis_reg(GRNdir,adata_RNA,adata_ATAC,genome,celltype,outdir,method)
```

#### *trans*-regulatory network
The output is 'cell_type_specific_trans_regulatory_{*celltype*}.txt', a matrix of the trans-regulatory score.
```python
LL_net.cell_type_specific_trans_reg(GRNdir,adata_RNA,celltype,outdir)
```
## Identify driver regulators by TF activity
### Instruction
TF activity, focusing on the DNA-binding component of TF proteins in the nucleus, is a more reliable metric than mRNA or whole protein expression for identifying driver regulators. Here, we employed LINGER inferred GRNs from sc-multiome data of a single individual. Assuming the GRN structure is consistent across individuals, we estimated TF activity using gene expression data alone. By comparing TF activity between cases and controls, we identified driver regulators. 

### Prepare
We need to *trans*-regulatory network, you can choose a network match you data best.
1. If your gene expression data are matched with cell population GRN, you can set
```python
network = 'cell population'
```
2. If your gene expression data are matched with certain cell type, you can set network to the name of this cell type.
```python
network = 'CD56 (bright) NK cells' # CD56 (bright) NK cells is the name of one cell type
```

### Calculate TF activity
The input is gene expression data, It could be the scRNA-seq data from the sc multiome data. It could be other sc or bulk RNA-seq data matches the GRN. The row of gene expresion data is gene, columns is sample and the value is read count (sc) or FPKM/RPKM (bulk).

```python
from LingerGRN.TF_activity import *
TF_activity=regulon(outdir,adata_RNA,GRNdir,network,genome)
```
Visualize the TF activity heatmap by cluster. If you want to save the heatmap to outdit, please set 'save=True'. The output is 'heatmap_activity.png'.
```python
save=True
heatmap_cluster(TF_activity,adata_RNA,save,outdir)
```
<div style="text-align: right">
  <img src="heatmap_activity_mm10.png" alt="Image" width="500">
</div>

### Identify driver regulator
We use t-test to find the differential TFs of a certain cell type by the activity. 
1. You can assign a certain cell type of the gene expression data by
```python
celltype='1'
```
2. Or, you can obtain the result for all cell types.
```python
celltype='all'
```

For example,

```python
celltype='1'
t_test_results=master_regulator(TF_activity,adata_RNA,celltype)
t_test_results
```

<div style="text-align: right">
  <img src="ttest_mm10.png" alt="Image" width="300">
</div>

Visulize the differential activity and expression. You can compare 2 different cell types and one cell type with other cell types. If you want to save the heatmap to outdit, please set 'save=True'. The output is 'box_plot'_+TFName+'_'+datatype+'_'+celltype1+'_'+celltype2+'.png'.

```python
TFName='Erg'
datatype='activity'
celltype1='1'
celltype2='Others'
save=True
box_comp(TFName,adata_RNA,celltype1,celltype2,datatype,TF_activity,save,outdir)
```

<div style="text-align: right">
  <img src="box_plot_Erg_activity_1_Others.png" alt="Image" width="300">
</div>

For gene expression data, the boxplot is:
```python
datatype='expression'
box_comp(TFName,adata_RNA,celltype1,celltype2,datatype,TF_activity,save,outdir)
```

<div style="text-align: right">
  <img src="box_plot_Erg_expression_1_Others.png" alt="Image" width="300">
</div>


